packages/workflow/src/astGrep/astGrep.ts

```diff
@@ -0,0 +1,114 @@
+import * as fs from "node:fs/promises";
+import { type NapiConfig, tsx as astGrepTsx } from "@ast-grep/napi";
+import { PLazy } from "../PLazy.js";
+import {
+  type AstGrepNodeContext,
+  astGrepNodeContext,
+  getFileContext,
+  getParentContext,
+} from "../contexts.js";
+import { wrapHelpers } from "../helpers.js";
+import { map } from "./map.js";
+import { replace } from "./replace.js";
+
+const astGrepHelpers = {
+  replace,
+  map,
+};
+
+type AstGrepHelpers = typeof astGrepHelpers;
+
+/**
+ *
+ * @param query Some description here
+ * @example
+ * ### Pass parameter as string
+ * ```ts
+ *   await astGrep("import React from 'react'");
+ * ```
+ * @example
+ * ```ts
+ *   await astGrep`import React from 'react'`;
+ * ```
+ * @example asdasd
+ * ### Pass parameter as NapiConfig [see ast-grep documentation](https://ast-grep.github.io/guide/rule-config.html)
+ * ```ts
+ *   await astGrep`import React from 'react'`;
+ * ```
+ */
+
+export function astGrep(
+  query: string | readonly string[] | NapiConfig,
+): PLazy<AstGrepHelpers> & AstGrepHelpers;
+/**
+ *
+ * @param query
+ * @param callback
+ * @example
+ * ```ts
+ *   await astGrep("import React from 'react'", ({ astGrep }) => {
+ *     astGrep("import React from 'react'");
+ *   });
+ * ```
+ */
+export function astGrep(
+  query: string | readonly string[] | NapiConfig,
+  callback: (helpers: AstGrepHelpers) => Promise<void> | void,
+): PLazy<AstGrepHelpers> & AstGrepHelpers;
+export function astGrep(
+  query: string | readonly string[] | NapiConfig,
+  callback?: (helpers: AstGrepHelpers) => Promise<void> | void,
+) {
+  const innerParentContext = getParentContext();
+
+  const grep =
+    typeof query === "object"
+      ? Array.isArray(query)
+        ? (query.join("") as string)
+        : (query as NapiConfig)
+      : query;
```

Comment:
if possible, try to avoid nested ternary and use imperative style, as it's much more readable